Sources/DistributedActors/Cluster/ClusterEventStream.swift

```diff
@@ -12,97 +12,195 @@
+        guard let shell = self.shell else { return }
+
+        await shell.whenLocal { __secretlyKnownToBeLocal in // TODO(distributed): this is annoying, we must track "known to be local" in typesystem instead
+            __secretlyKnownToBeLocal.unsubscribe(ref)
+        }
+    }
+
+    private func subscribe(_ oid: ObjectIdentifier, eventHandler: @escaping (Cluster.Event) -> Void) async {
+        guard let shell = self.shell else { return }
+
+        await shell.whenLocal { __secretlyKnownToBeLocal in // TODO(distributed): this is annoying, we must track "known to be local" in typesystem instead
+            __secretlyKnownToBeLocal.subscribe(oid, eventHandler: eventHandler)
+        }
+    }
+
+    private func unsubscribe(_ oid: ObjectIdentifier) async {
+        guard let shell = self.shell else { return }
+
+        await shell.whenLocal { __secretlyKnownToBeLocal in // TODO(distributed): this is annoying, we must track "known to be local" in typesystem instead
+            __secretlyKnownToBeLocal.unsubscribe(oid)
+        }
+    }
+
+    func publish(_ event: Cluster.Event, file: String = #filePath, line: UInt = #line) async {
+        guard let shell = self.shell else { return }
+
+        await shell.whenLocal { __secretlyKnownToBeLocal in // TODO(distributed): this is annoying, we must track "known to be local" in typesystem instead
+            __secretlyKnownToBeLocal.publish(event)
+        }
+    }
+
+    public func makeAsyncIterator() -> AsyncIterator {
+        AsyncIterator(self)
+    }
+
+    public class AsyncIterator: AsyncIteratorProtocol {
+        var underlying: AsyncStream<Cluster.Event>.Iterator!
 
-                return behavior.receiveSpecificSignal(_Signals.Terminated.self) { context, signal in
-                    if subscribers.removeValue(forKey: signal.id) != nil {
-                        context.log.trace("Removed subscriber [\(signal.id)], because it terminated")
-                    } else {
-                        context.log.warning("Received unexpected termination signal for non-subscriber [\(signal.id)]")
+        init(_ eventStream: ClusterEventStream) {
+            let id = ObjectIdentifier(self)
+            self.underlying = AsyncStream<Cluster.Event> { continuation in
+                Task {
+                    await eventStream.subscribe(id) { event in
+                        continuation.yield(event)
                     }
+                }
 
-                    return .same
+                continuation.onTermination = { _ in
+                    Task {
+                        await eventStream.unsubscribe(id)
+                    }
                 }
+            }.makeAsyncIterator()
+        }
+
+        public func next() async -> Cluster.Event? {
+            await self.underlying.next()
+        }
+    }
+}
+
+// FIXME(distributed): the only reason this actor is distributed is because of LifecycleWatch
+internal distributed actor ClusterEventStreamShell: LifecycleWatch {
```

Comment:
Note on the side / not urgent but FYI; we should not keep using the "shell" terms like that, the "shell" naming was "an actor around an instance that actually has all the logic" so we had `Thing` and `ThingShell` most of the time.

So I guess just cluster eventstream actor or something hm

Reply:
Renaming to `ClusterEventStreamActor`